sqlmesh/core/snapshot/evaluator.py

```diff
@@ -1669,6 +1677,114 @@ def get_custom_materialization_type(name: str) -> t.Type[CustomMaterialization]:
     return strategy_type
 
 
+class EngineManagedStrategy(MaterializableStrategy):
+    def create(
+        self,
+        table_name: str,
+        model: Model,
+        is_table_deployable: bool,
+        **render_kwargs: t.Any,
+    ) -> None:
+        is_snapshot_deployable: bool = render_kwargs["is_snapshot_deployable"]
+
+        if is_table_deployable and is_snapshot_deployable:
+            # We could deploy this to prod; create a proper managed table
+            logger.info("Creating managed table: %s", table_name)
+            self.adapter.create_managed_table(
+                table_name=table_name,
+                query=model.render_query_or_raise(**render_kwargs),
+                columns_to_types=model.columns_to_types,
+                partitioned_by=model.partitioned_by,
+                clustered_by=model.clustered_by,
+                table_properties=model.physical_properties,
+                table_description=model.description,
+                column_descriptions=model.column_descriptions,
+            )
+        else:
+            # Otherwise, fallback to the default logic of creating normal tables
+            super().create(
+                table_name=table_name,
+                model=model,
+                is_table_deployable=is_table_deployable,
+                **render_kwargs,
+            )
+
+    def insert(
+        self,
+        table_name: str,
+        query_or_df: QueryOrDF,
+        model: Model,
+        is_first_insert: bool,
+        **kwargs: t.Any,
+    ) -> None:
+        deployability_index: DeployabilityIndex = kwargs["deployability_index"]
+        snapshot: Snapshot = kwargs["snapshot"]
+        is_snapshot_deployable = deployability_index.is_deployable(snapshot)
+
+        if is_first_insert and is_snapshot_deployable and self.adapter.table_exists(table_name):
+            table_name_parsed = exp.to_table(table_name)
+
+            table_metadata = self.adapter.get_data_objects(
+                table_name_parsed.db, {table_name_parsed.name}
+            )
+            if len(table_metadata) == 0:
+                raise SQLMeshError(
+                    f"Table {table_name} exists but isnt being returned in the metadata. Cannot determine if it's a managed table or not"
+                )
+
+            if table_metadata[0].type != DataObjectType.MANAGED_TABLE:
```

Comment:
This is where the "forward-only in a virtual environment that might eventually get deployed" edge case is resolved.

I added an integration test `test_managed_model_upstream_forward_only` to exercise this code

Reply:
I'm not sure I understand how this scenario can happen.

Reply:
To summarise: it was due to unecessarily creating the "main" table next to the dev table, when we only needed to create the dev table